File: src/main/java/game/Game.java
package game;

import java.awt.BorderLayout;
import java.awt.Dimension;
import java.awt.Font;
import java.util.List;

import javax.swing.JFrame;
import ai.AI;

public class Game {
	private static final String WINDOW_NAME = "Minesweeper";
	private static final Font FONT = new Font("Calibri", Font.PLAIN, 20);
	private static final Dimension SCREEN_SIZE = new Dimension(700, 500);
	private static final int FRAME_TICK = 50;
	private static final int MOVE_TICK = 0;
	
	private final GameController controller;
	private AI ai;
	private Thread runThread;
	private Thread paintThread;
	private FieldPanel fieldPanel;
	private StatPanel statPanel;
	private GameMouseListener gameMouse;
	private Field field;
	private FieldView fieldView;
	private long startTime;
	private long worstTime = 0;
	private int gamesComplete = 0;
	private int gamesWon = 0;
	
	public Game(GameController controller, GameConfig config) {
		this.controller = controller;
		field = new Field(config, controller == GameController.AI);
		JFrame frame = new JFrame(WINDOW_NAME);
		frame.setSize(SCREEN_SIZE.width, SCREEN_SIZE.height);
		frame.setResizable(true);
		frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
		frame.setLocationRelativeTo(null);
		frame.setLayout(new BorderLayout());
		frame.setFont(FONT);
		if (controller == GameController.AI) {
			fieldView = new FieldView(config.getWidth(), config.getHeight());
			ai = new AI(config.getWidth(), config.getHeight(), fieldView);
		}
		fieldPanel = new FieldPanel(field, fieldView);
		gameMouse = new GameMouseListener(fieldPanel, config);
		fieldPanel.addMouseListener(gameMouse);
        frame.add(fieldPanel);
        statPanel = new StatPanel(field, this);
        frame.add(statPanel, BorderLayout.NORTH);
        frame.setVisible(true);
        startTime = System.currentTimeMillis();
	}
	
	public synchronized void start() {
		startPaint();
		runThread = new Thread("runThread") {
			public void run() {
				long nextTime = System.currentTimeMillis() + MOVE_TICK;
				while (!interrupted()) {
					try {
						sleep(Math.max(nextTime - System.currentTimeMillis(), 0));
					} catch (InterruptedException e) {
						break;
					}
					move();
					nextTime += MOVE_TICK;
				}
			}
		};
		runThread.start();
	}
	
	public synchronized void startPaint() {
		paintThread = new Thread("paintThread") {
			@Override
			public void run() {
				long nextTime = System.currentTimeMillis() + FRAME_TICK;
				while (!interrupted()) {
					try {
						sleep(Math.max(nextTime - System.currentTimeMillis(), 0));
					} catch (InterruptedException e) {
						break;
					}
					fieldPanel.repaint();
					statPanel.repaint();
					nextTime += FRAME_TICK;
				}
			}
		};
		paintThread.start();
	}

	public void move() {
		List<Click> clicks = null;

		if (controller == GameController.Player) {
			clicks = gameMouse.getClicks();
		} else if (controller == GameController.AI) {
			clicks = ai.solve(field.getBoard(), field.getMinesLeft());
		}

		if (clicks != null && !clicks.isEmpty()) {
			for (Click c : clicks) {
				if (c.getClickType() == ClickType.REVEAL) {
					field.changeBlock(c.getPoint());
				} else if (c.getClickType() == ClickType.FLAG) {
					field.flagBlock(c.getPoint());
				}
			}
			checkDone();
		}
    }

	public String getScore() {
		return String.format("Win Rate %.3f%%  [%,d  /  %,d]",
				100.0 * gamesWon / gamesComplete,
				gamesWon, gamesComplete);
	}
	
	private void checkDone() {
		if (field.checkWon()) {
			gamesWon++;
			reset();
		}
		else if (field.checkLost()) {
			reset();
		}
	}

	private void reset() {
		gamesComplete++;
		long totalTime = System.currentTimeMillis() - startTime;
		if (totalTime > worstTime) {
			worstTime = totalTime;
			field.save(totalTime);
		}

//		try {
//			Thread.sleep(1000);
//		} catch (InterruptedException e) {
//			e.printStackTrace();
//		}

		startTime = System.currentTimeMillis();
		field.reset();
		if (controller == GameController.AI) {
			ai.reset();
		}
	}

	public long getTime() {
		return System.currentTimeMillis() - startTime;
	}
}

File: src/main/kotlin/game/Game.kt
package game

import Main
import ai.AI
import ai.AIView
import game.ClickType.FLAG
import game.ClickType.REVEAL
import kotlinx.coroutines.delay
import kotlinx.coroutines.runBlocking
import java.awt.BorderLayout
import java.awt.Dimension
import java.awt.Font
import javax.swing.JFrame
import kotlin.concurrent.thread

class Game(
	private val controller: GameController,
	config: GameConfig
) {
	companion object {
		private val FONT = Font("Calibri", Font.PLAIN, 20)
		private val SCREEN_SIZE = Dimension(700, 500)
		private const val FRAME_TICK = 5000
		private const val MOVE_TICK = 0
	}

	private var aiView: AIView? = null
	private var ai: AI? = null

	private val field: Field = Field(config, controller == GameController.AI)
	private val fieldPanel: FieldPanel
	private val statPanel: StatPanel

	private val gameMouse: GameMouseListener

	private var startTime: Long

	private var gamesComplete = 0L
	private var gamesWon = 0L

	val time: Long
		get() = System.currentTimeMillis() - startTime

	val score: String
		get() = String.format(
			"Win Rate %.3f%%  [%,d  /  %,d]",
			100.0 * gamesWon / gamesComplete,
			gamesWon, gamesComplete
		)

	init {

		if (controller == GameController.AI) {
			aiView = AIView(config)
			ai = AI(config.width, config.height, aiView)
		}

		fieldPanel = FieldPanel(field, aiView)
		gameMouse = GameMouseListener(fieldPanel, config)
		fieldPanel.addMouseListener(gameMouse)

		statPanel = StatPanel(field, this)

		JFrame(Main.WINDOW_NAME).apply {
			setSize(SCREEN_SIZE.width, SCREEN_SIZE.height)
			isResizable = true
			defaultCloseOperation = JFrame.EXIT_ON_CLOSE
			setLocationRelativeTo(null)
			layout = BorderLayout()
			font = FONT
			add(fieldPanel)
			add(statPanel, BorderLayout.NORTH)
			isVisible = true
		}

		startTime = System.currentTimeMillis()
	}

	fun start() {
		thread(name = "paintThread") {
			runBlocking {
				var nextTime = System.currentTimeMillis() + FRAME_TICK
				while (true) {
					delay(nextTime - System.currentTimeMillis())
					fieldPanel.repaint()
					statPanel.repaint()
					nextTime += FRAME_TICK.toLong()
				}
			}
		}
		thread(name = "runThread") {
			runBlocking {
				var nextTime = System.currentTimeMillis() + MOVE_TICK
				while (true) {
					delay(nextTime - System.currentTimeMillis())
					move()
					nextTime += MOVE_TICK.toLong()
				}
			}
		}
	}

	private fun move() {
		val clicks = when (controller) {
			GameController.Player -> gameMouse.getClicks()
			GameController.AI -> ai?.solve(field.board, field.getMinesLeft())
		}
		if (clicks != null && clicks.isNotEmpty()) {
			clicks.forEach { (clickType, point) ->
				when (clickType) {
					REVEAL -> field.changeBlock(point)
					FLAG -> field.flagBlock(point)
				}
			}
			checkDone()
		}
	}

	private fun checkDone() {
		if (field.checkWon()) {
			gamesWon++
			reset()
		} else if (field.checkLost()) {
			reset()
		}
	}

	private fun reset() {
		gamesComplete++

		// try {
		// 	Thread.sleep(1000)
		// } catch (e: InterruptedException) {
		// 	e.printStackTrace()
		// }

		field.reset()
		startTime = System.currentTimeMillis()
		if (controller == GameController.AI) {
			ai?.reset()
		}
	}
}